compiler/rustc_middle/src/dep_graph/dep_node.rs

```diff
@@ -103,79 +197,151 @@ macro_rules! contains_eval_always_attr {
     ($($attr:ident $(($($attr_args:tt)*))* ),*) => ({$(is_eval_always_attr!($attr) | )* false});
 }
 
+#[allow(non_upper_case_globals)]
+pub mod dep_kind {
+    use super::*;
+    use crate::ty::query::{queries, query_keys};
+    use rustc_query_system::query::{force_query, QueryDescription};
+
+    // We use this for most things when incr. comp. is turned off.
+    pub const Null: DepKindStruct = DepKindStruct {
+        has_params: false,
+        is_anon: false,
+        is_eval_always: false,
+
+        can_reconstruct_query_key: || true,
+        force_from_dep_node: |_, dep_node| bug!("force_from_dep_node: encountered {:?}", dep_node),
+        try_load_from_on_disk_cache: |_, _| {},
+    };
+
+    // Represents metadata from an extern crate.
+    pub const CrateMetadata: DepKindStruct = DepKindStruct {
+        has_params: true,
+        is_anon: false,
+        is_eval_always: true,
+
+        can_reconstruct_query_key: || true,
+        force_from_dep_node: |_, dep_node| bug!("force_from_dep_node: encountered {:?}", dep_node),
+        try_load_from_on_disk_cache: |_, _| {},
+    };
+
+    pub const TraitSelect: DepKindStruct = DepKindStruct {
+        has_params: false,
+        is_anon: true,
+        is_eval_always: false,
+
+        can_reconstruct_query_key: || true,
+        force_from_dep_node: |_, _| false,
```

Comment:
`TraitSelect` and `CompileCodegenUnit` used to `bug!` on `force_from_dep_node`, as with `Null` and `CrateMetadata`. Was this change intentional? Sorry if this is obvious--I haven't read the whole PR.

Reply:
Actually, `force_from_dep_node` bailed out earlier (line 167), since those DepKinds are either anonymous, or with unreconstructible key. I was aiming at keeping the exact same behaviour, but panicking should be fine.

Reply:
Gotcha, I see now.